src/components/stepper/stepper.ts

```diff
@@ -0,0 +1,122 @@
+import {
+  Component,
+  Input,
+  ViewEncapsulation,
+  ContentChild,
+  Renderer,
+  ElementRef
+} from '@angular/core';
+import { BooleanFieldValue } from '@angular2-material/core/annotations/field-value';
+import { MdIcon, MdIconRegistry } from '@angular2-material/icon/icon';
+import { NgFormModel } from '@angular/common';
+
+@Component({
+  selector: 'md-stepper',
+  templateUrl: 'components/stepper/stepper.html',
+  styleUrls: ['components/stepper/stepper.css'],
+  providers: [MdIconRegistry],
+  directives: [MdIcon],
+  encapsulation: ViewEncapsulation.None
+})
+export class MdStepper {
+
+  steps: MdStep[] = [];
+
+  @Input() mode: 'linear' | 'nonlinear' = 'nonlinear';
+
+  addStep(step: MdStep): void {
+    if (this.steps.length === 0) step.active = true;
+    this.steps.push(step);
+  }
+
+  selectStep(step: MdStep): void {
+    this.currentStep.completed = this.currentStep.valid ? true : false;
+    if (this.mode === 'linear') {
+      if (!this.currentStep.valid) return;
+      //Checks if all steps before are completed or optional
+      for (var i = this.steps.indexOf(this.currentStep); i < this.steps.indexOf(step); i++) {
+        if (!this.steps[i].completed && !this.steps[i].optional) return;
+      }
+    }
+    if (step.completed && !step.editable) return;
+    this.steps.forEach((step) => {
+      step.active = false;
+    });
+    step.active = true;
+  }
+
+  get currentStep(): MdStep {
+    return this.steps.filter(step => step.active)[0];
+  }
+
+  selectNextStep(step: MdStep): void {
+    if (this.getNextEditableStep(step)) this.selectStep(this.getNextEditableStep(step));
+  }
+
+  selectPreviousStep(step: MdStep): void {
+    if (this.getPreviousEditableStep(step)) this.selectStep(this.getPreviousEditableStep(step));
+  }
+
+
+  getNextEditableStep(step: MdStep): MdStep {
+    if (this.steps.indexOf(this.currentStep) === this.steps.length - 1) return undefined;
+    for (var i = this.steps.indexOf(this.currentStep) + 1; i < this.steps.length; i++) {
+      if (this.steps[i].isTargetable) {
+        return this.steps[i];
+      }
+    }
+    return undefined;
+  }
+
+  getPreviousEditableStep(step: MdStep): MdStep {
+    if (this.steps.indexOf(this.currentStep) === 0) return undefined;
+    for (var i = this.steps.indexOf(this.currentStep) - 1; i >= 0; i--) {
```

Comment:
I think, since those functions (`previous` and `next`) are having a really close logic, we should be able to group this into a single function, which is more generic.